File: Python/mapsegments.py
import numpy as np
import math

from vector import *
from plane import *
from router import *

# =================================================================================================

class CSegmentPathNode:
    def __init__ (self, id = -1, nodePos = CVector (0,0,0), distToCenter = 0):
        self.id = id
        self.nodePos = nodePos
        self.distToCenter = distToCenter    # distance to segment center

    
class CSegmentPathEdge:
    def __init__ (self, segmentId = -1, startPos = CVector (0,0,0), endPos = CVector (0,0,0), distance = -1):
        self.segmentId = segmentId
        self.startPos = startPos
        self.endPos = endPos
        self.distance = distance


class CRouteData:
    def __init__ (self, distance, startPos, endPos):
        self.distance = distance
        self.startPos = startPos
        self.endPos = endPos

# =================================================================================================

class CWall (CPlane):
    def __init__ (self, vertices, position, isBoundary = False):
        super ().__init__ (vertices)
        self.position = position
        self.isBoundary = isBoundary


    def Translate (self, t):
        self.center += t

# =================================================================================================
# Segment class for map segments

class CMapSegment:
    def __init__ (self, x, y, id = -1):
        self.connected = [False, False, False, False]   # orthogonal neighbours
        self.connections = 0
        self.walls = []                                 # list of walls around the segment
        self.pathNodes = []                             # reference coordinates for los (line of sight) testing
        self.edgeIds = []                               # indices into list of other segments a los exists to
        self.position = (x, y)                          # position in the segment grid
        self.center = CVector (0,0,0)
        self.id = id
        self.actorCount = 0


    def SetId (self, id):
        self.id = id


    def GetId (self):
        return self.id


    def IsConnected (self, direction):
        return self.connected [direction] 


    def AddPathNode (self, node, distToCenter):
        self.pathNodes.append (CSegmentPathNode (self.GetId () * 10 + len (self.pathNodes) + 1, node, distToCenter))


    def FindPathEdge (self, edgeList, segmentId):
        for i in self.edgeIds:
            e = edgeList [i]
            if (e.segmentId == segmentId):
                return e
        return None


    def AddPathEdge (self, edgeList, edgeId, segmentId, startPos, endPos, distance):
        e = self.FindPathEdge (edgeList, segmentId)
        if (e is None):
            self.edgeIds.append (edgeId)
            return CSegmentPathEdge (segmentId, startPos, endPos, distance), True
        if (int (distance) < e.distance):
            e.startPos = startPos
            e.endPos = endPos
            e.distance = int (distance)
            return e, False
        return None, False

# =================================================================================================
# rectangular (2D) map of all segments

class CSegmentMap:
    def __init__ (self, scale, distanceQuality = 0):
        self.segments = None
        self.edgeList = []
        self.height = 0
        self.width = 0
        self.size = 0
        self.scale = scale
        self.distanceScale = 1000
        self.distanceQuality = distanceQuality
        self.router = None


    def __getitem__ (self, index):
        return self.segments [index // self.width, index % self.width]


    def __setitem__ (self, index, value):
        self.segments [index // self.width, index % self.width] = value


    def Create (self, width, height):
        self.width = width // 2
        self.height = height // 2
        self.size = self.width * self.height
        self.router = CRouter ()
        w = self.width * self.scale
        h = self.height * self.scale
        self.segments = np.full ([self.height, self.width], 0, CMapSegment)


    def AddSegment (self, x, y):
        s = CMapSegment (x, y, y * self.width + x)
        self.segments [y,x] = s
        return s


    def GetSegment (self, x, y):
        if (x < 0) or (y < 0) or (x >= self.width) or (y >= self.height):
            return None
        return self.segments [y,x]


    def GetSegmentById (self, id):
        return self.GetSegment (id % self.width, id // self.width)


    def SegPosFromId (self, id):
        return id % self.width, id // self.width


    def FindWall (self, walls, position):
        for wall in walls:
            if (wall.position == position):
                return wall
        return None


    # link current segment with neighbar at map position relative to the segment's position
    # direction: Orthogonal direction from current to adjacent segment
    # Add wall in direction if there is a wall, otherwise create link information in that 
    # direction (grid position of the adjacent segment)
    def LinkSegments (self, stringMap, segment, x, y, dx, dy, walls, direction):
        if (stringMap [y + dy][x + dx] != ' '): # i.e. there is a wall in that direction
            # so find wall by its position and append it to wall list
            segment.walls.append (self.FindWall (walls, (x + dx, y + dy)))  
            return 0
        else:   # no wall in direction direction
            # enter information about the adjacent segment's grid position
            # note: That segment will be or has been treated here as well, setting up its corresponding link information
            segment.connected [direction] = True
            segment.connections += 1
            return 1


    # Build the complete segment grid, determining walls and reachable neighbours around each segment
    def Build (self, map):
        rows = len (map.stringMap)
        cols = len (map.stringMap [0]) - 1
        self.Create (cols, rows)
        for y in range (1, rows, 2):
            for x in range (1, cols, 2):
                segment = self.AddSegment (x // 2, y // 2)
                self.LinkSegments (map.stringMap, segment, x, y, -1,  0, map.walls, 0)
                self.LinkSegments (map.stringMap, segment, x, y,  0, -1, map.walls, 1)
                self.LinkSegments (map.stringMap, segment, x, y,  1,  0, map.walls, 2)
                self.LinkSegments (map.stringMap, segment, x, y,  0,  1, map.walls, 3)
        self.CreatePathNodes (map)
        if (self.distanceQuality == 1):
            self.distanceScale = 1000
            while (not self.CreatePathEdges (map)):
                self.ResetPathData ()
            self.CreateDistanceTable ()


    def SegmentCenter (self, x, y, scale):
        return CVector ((x + 0.5) * scale, scale / 4, -(self.height - y - 0.5) * scale)


    def ComputeSegmentCenters (self, map):
        for y, segmentRow in enumerate (self.segments):
            for x, segment in enumerate (segmentRow):
                segment.center = self.SegmentCenter (x, y, map.scale)


    # create four 2D coordinates for each segment: edge centers
    # outer coordinates are slightly offset into the segment
    # Purpose: Finding out to which other segments a given segment has a los (line of sight)
    # This will be used in distance calculation for sound sources create (pseudo) positional sound
    # los coordinates are only created if there is no wall at their edge
    def CreatePathNodes (self, map):
        offsets = [(-1,0), (0,-1), (1,0), (0,1)]    # [(-1,0), (0,-1), (1,0), (0,1), (0,0), (-1,-1), (1,-1), (1,1), (-1,1)]
        nodeCount = 0
        for y, segmentRow in enumerate (self.segments):
            radius = map.scale / 2
            for x, segment in enumerate (segmentRow):
                segment.center = self.SegmentCenter (x, y, map.scale)
                for direction, o in enumerate (offsets):
                    if (segment.connected [direction]):
                        nodeOffset = CVector (radius * o [0], 0, radius * o [1])
                        segment.AddPathNode (segment.center + nodeOffset, nodeOffset.Length ())
                        nodeCount += 1
        print ("created " + str (nodeCount) + " path nodes")


    def ResetPathData (self):
        emptyList = []
        for i in range (self.size - 1):
            self [i].edgeIds = emptyList [:]
        self.edgeList = emptyList [:]

    # CreatePathEdges casts a ray from each path nodes of each segment to each path node of each other
    # segment. When such a ray does not intersect any interor walls of the map, there is a line of sight
    # between the two segments. Rays will not be cast to path nodes behind the current path node as seen
    # from the path node's segment's center. That direction will be handled by the segment's path node at
    # the opposite segment edge.
    def CreatePathEdges (self, map):    
        losTests = 0
        edgeCount = 0
        lMax = 0
        for i in range (self.size - 1):
            si = self [i]
            for j in range (i + 1, self.size):
                sj = self [j]
                for ni in si.pathNodes:
                    n = ni.nodePos - si.center
                    for nj in sj.pathNodes:
                        v = nj.nodePos - ni.nodePos
                        d = v.Length ()
                        if (d == 0):
                            haveLoS = True
                        else:
                            if (n.Dot (v) < 0.0):
                                continue
                            haveLoS = True
                            for w in map.walls:
                                if w.isBoundary:   # don't consider walls surrounding the map as there is no segment behind them
                                    continue
                                losTests += 1
                                vi, penetrated = w.LineIntersection ([ni.nodePos, nj.nodePos])
                                if (vi is None):   # vector from path nodes ni to nj crosses a wall
                                    continue
                                if w.Contains (vi):
                                    haveLoS = False
                                    break   # no need for further tests, we don't have LoS anymore already
                        if (haveLoS):
                            d += ni.distToCenter + nj.distToCenter
                            l = int (math.ceil (d * self.distanceScale))
                            e, isNewEdge = si.AddPathEdge (self.edgeList, edgeCount, sj.id, ni.nodePos, nj.nodePos, d)
                            if isNewEdge:
                                self.edgeList.append (e)
                                edgeCount += 1
                            e, isNewEdge = sj.AddPathEdge (self.edgeList, edgeCount, si.id, nj.nodePos, ni.nodePos, d)
                            if isNewEdge:
                                self.edgeList.append (e)
                                edgeCount += 1

            if (lMax > self.router.maxCost):
                m_distanceScale = int (self.router.maxCost * self.distanceScale) // lMax
                return False

            self.edgeTable = np.empty ([edgeCount], CSegmentPathEdge)
            for i, e in enumerate (self.edgeList):
                self.edgeTable [i] = e
            return True
        # diagnostic message for test purposes
        # wallCount = [0, 0]
        # for w in map.walls:
        #     wallCount [w.isBoundary] += 1
        # print ("performed {0} LoS tests for {1} walls ({2} boundary walls were ignored). Created {3} edges.".format (losTests, wallCount [0], wallCount [1], edgeCount))


    # build a table with distances from each segment to each other reachable segment
    # For start and end segment, subtract the distance between the segment centers and the segments' edge center nodes used in
    # the shortest path between these segment. When determining the distance between two actors, the distances of the
    # actors to the edge center nodes of their respective segment. This yields an acceptable approximation of the 
    # path distance between the two actors.
    # 
    # +---------+---------+---------+---------+
    # |         |         |         |S4       | 
    # |      /--x---------x---------x-\__     | 
    # |     /   |         |         |    \-B  | 
    # +----x----+---------+---------+---------+
    # |S2  |    |
    # |    |    |
    # |    |    |
    # +----x----+
    # |S1   \   |
    # |      \  |
    # |       A |
    # +---------+
    
    def CreateDistanceTable (self):
        self.router.Create (self.size)
        self.distanceTable = np.full ([self.size, self.size], 0, CRouteData)
        for i in range (self.size - 1):
            si = self [i]
            self.router.FindPath (si.id, -1, self)
            for j in range (i + 1, self.size):
                route = self.router.BuildRoute (j)
                if (len (route) < 3):
                    self.distanceTable [i,j] = \
                    self.distanceTable [j,i] = CRouteData (-1, CVector (0,0,0), CVector (0,0,0))
                else:
                    p0 = self.edgeTable [route [1].edge].startPos
                    p1 = self.edgeTable [route [-2].edge].startPos
                    x, y = self.SegPosFromId (i)
                    v0 = p0 - self.segments [y,x].center
                    x, y = self.SegPosFromId (j)
                    v1 = p1 - self.segments [y,x].center
                    cost = self.router.FinalCost (j) / self.distanceScale - v0.Length () - v1.Length ()
                    self.distanceTable [i,j] = CRouteData (cost, p0, p1)
                    self.distanceTable [j,i] = CRouteData (cost, p1, p0)
        self.router.Destroy ()


    # reset the actor count in each segment that had previously been computed
    def ResetActorCounts (self):
        for row in self.segments:
            for col in row:
                col.actorCount = 0

    # increase the actor count in the segment at map position x,y
    def CountActorAt (self, x, y):
        self.segments [y,x].actorCount += 1


    def ActorCountAt (self, x, y):
        return self.segments [y,x].actorCount


    def Distance (self, i, j):
        return self.distanceTable [i,j]

# =================================================================================================

File: CPP/mapsegments.cpp
#include "mapsegments.h"
#include "router.h"
#include "plane.h"

// =================================================================================================
// Segment class for map segments

CMapSegment::CMapSegment(int x, int y, int id) {
    m_connected[0] =
    m_connected[1] =
    m_connected[2] =
    m_connected[3] = false;
    m_connections = 0;
    m_position = { x, y };
    m_center = CVector(0, 0, 0);
    m_id = id;
    m_actorCount = 0;
}


CSegmentPathEdge* CMapSegment::FindPathEdge(CList<CSegmentPathEdge>& edgeList, int segmentId) {
    for (auto [i, edgeId] : m_pathEdgeIds) {
        CSegmentPathEdge& e = edgeList[edgeId];
        if (e.m_segmentId == segmentId)
            return &e;
    }
    return nullptr;
}


bool CMapSegment::AddPathEdge(CList<CSegmentPathEdge>& edgeList, int edgeId, int segmentId, CVector& startPos, CVector& endPos, int distance, CSegmentPathEdge& edge) {
    CSegmentPathEdge* e = FindPathEdge(edgeList, segmentId);
    if (!e) {
        m_pathEdgeIds.Append(edgeId);
        edge = CSegmentPathEdge(segmentId, startPos, endPos, distance);
        return true;
    }
    if (distance < e->m_distance) {
        e->m_startPos = startPos;
        e->m_endPos = endPos;
        e->m_distance = int(distance);
        return false;
    }
    return false;
}

// =================================================================================================
// rectangular (2D) map of all segments

void CSegmentMap::Create(int width, int height) {
    Destroy();
    m_width = width / 2;
    m_height = height / 2;
    m_size = m_width * m_height;
    int w = int (m_width * m_scale);
    int h = int (m_height * m_scale);
    m_segments.Create(height);
    for (auto row : m_segments)
        row->Create(width);
}


void CSegmentMap::Destroy(void) {
    m_pathEdgeList.Destroy();
    m_pathEdgeTable.Destroy();
    m_distanceTable.Destroy ();
    m_segments.Destroy();
    m_height = 
    m_width = 
    m_size = 0;
}


CMapSegment* CSegmentMap::AddSegment(int x, int y) {
    CMapSegment* s = &m_segments[y][x];
    *s = CMapSegment(x, y, y * m_width + x);
    return s;
}


CWall* CSegmentMap::FindWall(CList<CWall>& walls, CMapPosition position) {
    for (auto [i, wall] : walls)
        if (wall.m_position == position)
            return &wall;
    return nullptr;
}


// link current segment with neighbar at map position relative to the segment's position
// direction: Orthogonal direction #include "current to adjacent segment
// Add wall in direction if there is a wall, otherwise create link information in that 
// direction (grid position of the adjacent segment)
int CSegmentMap::LinkSegments(CList<CString>& stringMap, CMapSegment* segment, int x, int y, int dx, int dy, CList<CWall>& walls, int direction) {
    if (stringMap[y + dy][x + dx] != ' ') { // imap.m_emap.m_ there is a wall in that direction
        // so find wall by its position and append it to wall list
        segment->m_walls.Append(FindWall(walls, CMapPosition(x + dx, y + dy)));
        return 0;
    }
    else {   // no wall in direction direction
        // enter information about the adjacent segment's grid position
        // note: That segment will be or has been treated here as well, setting up its corresponding link information
        segment->m_connected[direction] = true;
        segment->m_connections++;
        return 1;
    }
}


// Build the complete segment grid, determining walls and reachable neighbours around each segment
void CSegmentMap::Build(CList<CString>& stringMap, CList<CWall>& walls, float scale) {
    int rows = int (stringMap.Length());
    int cols = int (stringMap[0].Length()) - 1;
    Create(cols, rows);
    for (int y = 1; y < rows; y += 2) {
        for (int x = 1; x < cols; x += 2) {
            CMapSegment* segment = AddSegment(x / 2, y / 2);
            LinkSegments(stringMap, segment, x, y, -1,  0, walls, 0);
            LinkSegments(stringMap, segment, x, y,  0, -1, walls, 1);
            LinkSegments(stringMap, segment, x, y,  1,  0, walls, 2);
            LinkSegments(stringMap, segment, x, y,  0,  1, walls, 3);
        }
    }
    CreatePathNodes(scale);
    if (m_distanceQuality == 1) {
        m_distanceScale = 1000;
        // if path edge lengths are too great for the router, CreatePathEdges will adjust the distance scale 
        // and then needs to be run again. This should only happen once.
        while (!CreatePathEdges (walls))
            ResetPathData ();
        CreateDistanceTable();
    }
}



void CSegmentMap::ComputeSegmentCenters(float scale) {
    for (int y = 0; y < m_height; y++)
        for (int x = 0; y < m_width; x++)
            m_segments[y][x].m_center = SegmentCenter(x, y, scale);
}


// create four 2D coordinates for each segment: edge centers
// outer coordinates are slightly offset into the segment
// Purpose: Finding out to which other segments a given segment has a los (line of sight)
// This will be used in distance calculation for sound sources create (pseudo) positional sound
// los coordinates are only created if there is no wall at their edge
void CSegmentMap::CreatePathNodes(float scale) {
    CMapPosition offsets[] = { CMapPosition(-1,0), CMapPosition(0,-1), CMapPosition(1,0), CMapPosition(0,1) };    // [(-1,0), (0,-1), (1,0), (0,1), (0,0), (-1,-1), (1,-1), (1,1), (-1,1)]
    int nodeCount = 0;
    int y = 0;
    for (auto segmentRow : m_segments) {
        float radius = scale / 2;
        int x = 0;
        for (auto segment : *segmentRow) {
            segment->m_center = SegmentCenter(x, y, scale);
            for (int direction = 0; direction < sizeofa(offsets); direction++) {
                if (segment->m_connected[direction]) {
                    CVector nodeOffset = CVector(radius * offsets[direction].m_x, 0, radius * offsets[direction].m_y);
                    segment->AddPathNode(segment->m_center + nodeOffset, nodeOffset.Len());
                    nodeCount++;
                }
            }
            x++;
        }
        y++;
    }
}


void CSegmentMap::ResetPathData (void) {
    for (int i = 0; i < m_size; i++)
        (*this) [i].m_pathEdgeIds.Destroy ();
    m_pathEdgeList.Destroy ();
}


// CreatePathEdges casts a ray #include "each path nodes of each segment to each path node of each other
// segmentmap.m_ When such a ray does not intersect any interor walls of the map, there is a line of sight
// between the two segmentsmap.m_ Rays will not be cast to path nodes behind the current path node as seen
// #include "the path node's segment's centermap.m_ That direction will be handled by the segment's path node at
// the opposite segment edgemap.m_
bool CSegmentMap::CreatePathEdges(CList<CWall>& walls) {
    router.Create (m_size);
    int edgeCount = 0;
    int lMax = 0;
    for (int i = 0; i < m_size - 1; i++) {
        CMapSegment& si = (*this)[i];
        for (int j = i + 1; j < m_size; j++) {
            CMapSegment& sj = (*this)[j];
            for (auto ri : si.m_pathNodes) {
                CSegmentPathNode& ni = ri.second;
                CVector n = ni.m_nodePos - si.m_center;
                for (auto rj : sj.m_pathNodes) {
                    CSegmentPathNode& nj = rj.second;
                    CVector v = nj.m_nodePos - ni.m_nodePos;
                    float d = v.Len();
                    bool haveLoS = false;
                    if (d == 0.0f)
                        haveLoS = true;
                    else {
                        if (n.Dot(v) < 0.0f)
                            continue;
                        haveLoS = true;
                        for (auto rw : walls) {
                            CWall& w = rw.second;
                            if (w.m_isBoundary)   // don't consider walls surrounding the map as there is no segment behind them
                                continue;
                            CVector vi;
                            w.LineIntersection(vi, ni.m_nodePos, nj.m_nodePos);
                            if (!vi.IsValid())   // vector #include "path nodes ni to nj crosses a wall
                                continue;
                            if (w.Contains(vi)) {
                                haveLoS = false;
                                break;   // no need for further tests, we don't have LoS anymore already
                            }
                        }
                    }
                    if (haveLoS) {
                        d += ni.m_distToCenter + nj.m_distToCenter;
                        int l = int (ceil (d * m_distanceScale));
                        if (lMax < l)
                            lMax = l;
                        CSegmentPathEdge e;
                        bool isNewEdge = si.AddPathEdge(m_pathEdgeList, edgeCount, sj.m_id, ni.m_nodePos, nj.m_nodePos, l, e);
                        if (isNewEdge) {
                            m_pathEdgeList.Append(e);
                            edgeCount++;
                        }
                        isNewEdge = sj.AddPathEdge(m_pathEdgeList, edgeCount, si.m_id, nj.m_nodePos, ni.m_nodePos, l, e);
                        if (isNewEdge) {
                            m_pathEdgeList.Append(e);
                        edgeCount++;
                        }
                    }
                }
            }
        }
    }
    if (lMax > router.MaxCost ()) { // max. permissible edge length in router
        m_distanceScale = router.MaxCost () * m_distanceScale / lMax;
        return false;
    }
    m_pathEdgeTable.Create(edgeCount);
    for (auto [i, e] : m_pathEdgeList)
        m_pathEdgeTable[i] = e;
    m_pathEdgeList.Destroy();
    return true;
}
                        

// build a table with distances #include "each segment to each other reachable segment
// For start and end segment, subtract the distance between the segment centers and the segments' edge center nodes used in
// the shortest path between these segmentmap.m_ When determining the distance between two actors, the distances of the
// actors to the edge center nodes of their respective segmentmap.m_ This yields an acceptable approximation of the 
// path distance between the two actorsmap.m_
// 
// +---------+---------+---------+---------+
// |         |         |         |S4       | 
// |      _/-x---------x---------x-\__     | 
// |     /   |         |         |    \-B  | 
// +----x----+---------+---------+---------+
// |S2  |    |
// |    |    |
// |    |    |
// +----x----+
// |S1   \   |
// |      \  |
// |       A |
// +---------+
    
void CSegmentMap::CreateDistanceTable(void) {
    m_distanceTable.Create(m_size);
    for (auto row : m_distanceTable)
        row->Create(m_size);
    for (int i = 0; i < m_size - 1; i++) {
        CMapSegment& si = (*this)[i];
        router.FindPath(si.m_id, -1, *this);
        for (int j = i + 1; j < m_size; j++) {
            CList<CRouteNode> route = router.BuildRoute(j);
            if (route.Length() < 3) {
                m_distanceTable[i][j] =
                m_distanceTable[j][i] = CRouteData(-1, CVector(0, 0, 0), CVector(0, 0, 0));
            }
            else {
                CVector& p0 = m_pathEdgeTable[route[1].m_edgeId].m_startPos;
                CVector& p1 = m_pathEdgeTable[route[-2].m_edgeId].m_startPos;
                auto sp0 = SegPosFromId(i);
                CVector v0 = p0 - m_segments[sp0.y][sp0.x].m_center;
                auto sp1 = SegPosFromId(j); 
                CVector v1 = p1 - m_segments[sp1.y][sp1.x].m_center;
                float cost = float (router.FinalCost (j)) / float (m_distanceScale) - v0.Len () - v1.Len ();
                m_distanceTable[i][j] = CRouteData(cost, p0, p1);
                m_distanceTable[j][i] = CRouteData(cost, p1, p0);
            }
        }
    }
    router.Destroy();
}


// reset the actor count in each segment that had previously been computed
void CSegmentMap::ResetActorCounts(void) {
    for (auto row : m_segments)
        for (auto col : *row)
            col->m_actorCount = 0;
}

// =================================================================================================
